### PR TITLE
Safe Mode Brick Option

### LINKED ISSUE #1039: Permanent OTA Safe Mode Feature
**Describe the problem you have/What new integration you would like**

ESPHome stores the boot attempts on flash instead of RTC and once ESPHome enters safe mode, it doesn't exit safe mode until it has been OTA'd or is disabled by the user. This would allow the boot attempts to persist though a power cycle. Effectively, I am looking for a safe mode "brick".

**Please describe your use case for this integration and alternatives you've tried:**

There are times where instead of a bootloop, ESPHome instead freezes from a lambda, resulting in it not being able to enter safe mode. It would be great if I would be able to unplug and plug in the device multiple times to force it to enter safe mode. Creating the "brick" would be great because I may not know exactly what iteration of "bootloop" ESPHome may be on. This would allow me to for sure catch it in safe mode if it was permanently activated.

**Additional context**

Since this feature requires storing data on the flash, it would need the build flag to be enabled. However, it seems that enabling `esp8266_restore_from_flash` currently allows all components that can write to flash, write their state to flash unless it is manually disabled. 

In this case, I would only want the boot counter to be stored to flash without anything else being written to limit the damage to the flash. I'm not too familiar with the code, but it seems that there would need to be a refactor for how configuration is currently written to flash in order to make this happen.

### PR DESCRIPTION
**Pull request by alexyao2015**
## Description:
Adds safe mode brick option. 
Permanently enters safe mode once in.

**WHY:**
Surefire way to enter safe mode by power cycling.
1) Powercycle to enter safe mode
2) Prevents excessive bootloop counter flash writes. (hit the max and it stops for good)
3) Prevents the "I don't know what bootloop count I'm on" issue
4) [Tasmota's `Fast Power Cycle Device Recovery`](https://tasmota.github.io/docs/Device-Recovery/#fast-power-cycle-device-recovery) has nearly identical behavior. 

**WHEN:**
1) While loop halting all functions including OTA.
2) Unforeseen bad lambdas. 

**Why BRICK???:**
- Prevents infinite bootlooping in and out of safe mode, destroying flash. 
- Solves child flipping switch/power supply issue from completely destroying the flash. 

**Why named BRICK??**
Refers to soft brick. Alerts user when viewing the config option and documentation to read it throughly. 

**Harms user experience??:**
It is defaulted to OFF. Very clear documentation is written explaining it. 

**Controversial?**
Suggest a better simpler solution. 

**Why not:**

- Alternate entering safe mode and no safe mode, resetting after x options, make configurable for user
Long, complex PR == BUGS
BUGS == PERMANENT HARD BRICK (Tuya devices) == LANDFILL

- NOT SOFT brick?
You have no idea what bootloop cycle you are on. On cycle 8? How will you know that you're on 8 or 1 when you can't see serial. Go over boot count and you start over again. 

**Related issue (if applicable):** fixes esphome/feature-requests#1039

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#890

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
So this basically adds option to store the boot loop counter in permanent flash instead of rtc mem.

How does this differ from ESP32 normal behavior? At least it should be explained in docs if this affects esp8266 only.

What do you mean by brick? for me this term is not adequate as it is at least ambiguous of what software is running (if any is running at all)
Could you do some test and check if ESP32 stays in Safe Mode even after a power cycle after entering safe mode? I'd bet it does.

**Comment by alexyao2015** [comment]
Hmm good questions. 

As to the esp32, I'm not sure if/what will happen or if the default behavior on the esp32 is to store the counter on flash. I don't currently own one to test that. Would you be able to potentially try it?

By brick, I mean the device is effectively rendered useless until it has been OTA'd where the preference will be reset. Again, I'm not sure about esp32 or if the bootloop counter will reset with an OTA. I do know it resets on the esp8266 after an OTA. The device will go into a state where it is permanently locked in safe mode and will not proceed further in boot regardless of a power cycle.

**Comment by alexyao2015** [comment]
The other thing this changes is it prevents the preference from being cleaned after entering safe mode if the setting is enabled. This is where the brick comes into play. Since the preference isn't being cleared, it will always enter safe mode once it enters it the first time.

**Comment by alexyao2015** [comment]
It occurred to me that this will still have an effect on esp32 in that it will still "brick" the device.

**Comment by glmnet** [comment]
I just read the code so far, to my understanding now this has no effect at all on ESP32. Will do some tests later

File: esphome/components/ota/ota_component.cpp [hunk 1]
```diff
@@ -355,18 +355,21 @@ void OTAComponent::set_auth_password(const std::string &password) { this->passwo
...
+  this->prefs_ = global_preferences.make_preference<uint32_t>(233825507UL, store_bootloops_in_flash_and_brick);
+  this->safe_mode_prefs_value_ = this->read_prefs_();
 
-  ESP_LOGCONFIG(TAG, "There have been %u suspected unsuccessful boot attempts.", this->safe_mode_rtc_value_);
+  ESP_LOGCONFIG(TAG, "There have been %u suspected unsuccessful boot attempts.", this->safe_mode_prefs_value_);
 
-  if (this->safe_mode_rtc_value_ >= num_attempts) {
-    this->clean_rtc();
+  if (this->safe_mode_prefs_value_ >= num_attempts) {
+    if (!store_bootloops_in_flash_and_brick) {
```
Comment (alexyao2015) [inline esphome/components/ota/ota_component.cpp:370]: This line is the only effect it will have on an esp32. This is where brick comes from.

**Comment by napieraj** [comment]
What is the purpose of bricking the device when we have a configurable safe mode OTA window? getting back into safe mode is trivial enough if you miss the first opportunity, no?

in keeping with the move in naming convention from `rtc` to `prefs`, `clean_rtc()` should also be renamed, as it stands there is ambiguity between `clean_rtc()` and the other functions which now use the prefs nomenclature

**Comment by alexyao2015** [comment]
Imagine this scenario. You write some lambda that when run, the whole esp freezes. This results in the ESP not bootlooping, but freezing, requiring a full power cycle to unfreeze. Due to a power cycle being required, the boot loop counter is lost. 

In the case of catching the safe mode, its extraordinarily hard to catch it if you're in the situation where the ESP does freeze. You may not know what boot loop counter you are currently on while you are manually power cycling it. Bricking it would ensure you would be able to access safe mode quickly without having to potentially cycle it hundreds of times and reduces frustration. Especially since it typically would never reach 10 cycles during normal operation. After the set time, the boot counter would be reset normally. 

And good catch, will rename that as well.

**Comment by glmnet** [comment]
How can you freeze the esp? The watchdog should reset it and enter in safe mode by itself

**Comment by alexyao2015** [comment]
Hmm I'm not sure then. I'm almost sure I've had times when it did freeze. Either way, this provides a more sure-fire way to ensure you can get into safe mode.

**Comment by alexyao2015** [comment]
Even with watchdog, you would normally need to sit there and wait x amount of time for the esp to reset 10 times. And once you wait that time, you only have a 5 minute interval to OTA it before you would be forced to wait that all over again. This lets you let the device reset on its own and you can come back to OTA it. Or you can forcibly induce it through power cycles. Its unlikely for it to reach 10 unhandled boots on its own anyways. If it does, you likely should OTA it anyway.

**Comment by napieraj** [comment]
> You may not know what boot loop counter you are currently on while you are manually power cycling it.

You do not have to "catch" it at exactly the number of failed `num_attempts`, all it takes is to cycle it 10-or-so times, it can be 15, it can be 20 (with a few seconds inbetween) and you can be reasonably certain it goes into safe mode. I see no need to semi-permanently brick the device and require and complete firmware reflash.

In short: the option to store failed boot loops in flash could be useful for some users, but bricking causes more harm than it brings benefit. Imagine the scenario in which a child is playing with the light switch controlling a smart bulb with the brick option enabled, or some kind of power fault/brownout that would cause the esp to temporarily bootloop and once again require a complete reflash to return to operation

**Comment by alexyao2015** [comment]
Then don't enable the feature for those devices? It's turned off by default and has a pretty detailed note explaining that it does brick the device in the docs as well as in the setting name.

And no you can't really be certain it's in safe mode. If say it's on bootloop 5 and you decide to power cycle it, you have no clue if it's on bootloop 5 or 1. So you're just sitting there praying and hoping you hit the magic number of powercycles. Hit the wrong number and you start over again. And you don't have a clue when you're starting over with the bootloop counter either.

**Comment by napieraj** [comment]
That's the issue, you're cramming two features into a single parameter -- what if I do want to store failed boot loops in flash but not risk bricking my devices?

we already have the `esp8266_restore_from_flash` param, would it not be sensible to use this value to determine where to store failed boot loops as well, and potentially add the brick option as its own option if someone really wants that?

**Comment by alexyao2015** [comment]
If you do toggle `esp8266_restore_from_flash`, it will automatically save the bootloops to flash same if you have an esp32. That's how esphome works underneath it all.

**Comment by napieraj** [comment]
This is not my understanding, the previous code calls `.make_preference` with the *in_flash* bool set to false, regardless of `esp8266_restore_from_flash` setting -- there is no check in the preference handler, see: https://esphome.io/api/preferences_8cpp_source.html#l00185

**Comment by alexyao2015** [comment]
Hmm I'll have to test that then. Pretty sure from my testing, it stored in flash anyways

**Review by badbadc0ffee — CHANGES_REQUESTED** [review]
Deliberately *bricking* a device does not make any sense.

At least the wording is very awkward and must be changed to something like *power_cycle_recovery* if that is what you are specifically aiming at. (That is also a common terminology as seen [here](https://tasmota.github.io/docs/Device-Recovery/#fast-power-cycle-device-recovery).)

Generally, I have the feeling that the *safe_mode*, which is already in place already does its job. I don't understand why another *bricked_mode* would be necessary or useful. I'd definitely *not* want to permanently lock the device in such a mode which could be reached by error as well: a timeout after which the system would try to return to normal mode would be indispensable - which leads us back to the existing *safe_mode*.

**Comment by alexyao2015** [comment]
[This](https://github.com/esphome/esphome/pull/1420#issuecomment-745717871) is a prime use case for the feature. Although a new suggestion is to force it into safe mode until it can cleanly exit after the default duration of 5 minutes. The Tasmota feature linked is effectively the same as it completely resets the entire device and forces a reconfiguration. In ESPHome, we don't have reconfiguration, but a device can be "reconfigured" through an OTA.

**Comment by alexyao2015** [comment]
Heres an update on some of the changes and reasons for a full-on permanent safe mode.

1) In order to preserve the flash, we must ultimately land the user permanently in safe mode. 
2) Enabling esp8266_restore_from_flash should only affect components that the user may adjust. Linking this to that setting would potentially allow a user to destroy their flash if it bootloops when that setting is on. Consider the device bootlooping for a day or more before a user realizes. The device would likely be toast by then.
3) I've now included full validation for all setting so an incompetent user cannot enter settings that will brick the device.
4) Alternating between safe mode/normal mode results in [too much code being changed for little benefit](https://github.com/alexyao2015/esphome/commit/28f27dfc752b79c05fe69b6d46bf624f27c0fec0). Ultimately, we still need to enter a permanent safe mode to preserve flash.
5) Should a user face issues with a child flicking a light switch, they may either choose to not enable this setting or simply enable the bootcounter to a number that would avoid this issue since OTA is now configurable. 
6) Since Tasmota has a feature similar to this as pointed out by @badbadc0ffee, it likely will not be a major impact.


Future Todos:

1) Perhaps make this the default behavior to enter safe mode unless deep sleep has been enabled. 
2) This should either be not allowed if deep sleep is enabled or a large warning should be placed in the documentation since it will write to flash on every wakeup. Not sure of which is better atm.
3) Rename feature to something else? Suggestions? IMO this resembles the definition of soft brick pretty well.

**Comment by badbadc0ffee** [comment]
I still find the wording very inapropriate.

And I want to point out that the Tasmota feature differs sustantially: it tries to avoid the necessity for firmware updates for the normal user by all means. Tasmota tries to help end users. This aims at developers where I'd be comfortable to accept when someone needs to open a device to reflash it because he shot himself into his knee by putting a blocking loop into an on_boot trigfer or similar things.

I do totally agree that this must be very thoroughly tested to avoid introducing bugs in this critical system phase. What I don't see yet is the automated test suite that would reflect that. ;-)

**Comment by alexyao2015** [comment]
The issue is here that we don't have a config. We have safe mode which is pretty similar to wiping the config. In safe mode, we sit and wait for a new config to come through OTA. Ultimately, I see this as an important last resort to catch any code that may have otherwise fully bricked the device, causing the user to have to discard the device (in my case since it was tuya) or have to open it up and reflash it. I've had to throw away two otherwise perfectly fine devices simply because they were bootlooping without a real way to enter safe mode.

There may always be bugs that cause a bootloop or freeze but we need a safety net to catch anything that falls through. @glmnet mentioned placing a while loop in a lambda that would 1) Freeze the device without tripping watchdog 2) block OTA since the device is stuck in a while loop. Point being is there may be more cases like this that we may not be able to think of now but need a failsafe for it. In this case, it's a purposefully bricking while loop, next time, it's a user that forgot to place break inside the while loop, then it's a user that bricks it in a way we cant foresee today. The fact that we have lambdas makes it possible to brick the device in all kinds of ways we cant foresee. 

The wording can easily be changed, but support for the feature is important.

**Comment by alexyao2015** [comment]
> And I want to point out that the Tasmota feature differs sustantially: it tries to avoid the necessity for firmware updates for the normal user by all means. 

If you use ESPHome, it's expected that you know how to OTA because there is no other way you're changing the configuration. You cannot say you use ESPHome without knowing how to do an OTA. Setting up for someone else who doesn't know how to OTA? Either 1) disable the feature beforehand like in Tasmota or 2) teach them how to OTA/reconfig like they would have to in Tasmota if the config got reset.

What's the point of avoiding an OTA besides it being annoying? I find it on the same level of annoying (perhaps a little less annoying) as having to re-setup the wifi, templates, and settings in Tasmota. There is no difference.

**Comment by glmnet** [comment]
There is one thing this attempt to fix:

Device doesn't reboot itself, it halts, this can be done with poor code, watchdog doesn't reset it and no services (OTA) is attended. 
Here we could reset the device with the reset button, this has two issues:
1. We don't know how many times we have to push the button to stay in safe mode, as if we overshot then it will start count again 
2. We might not have access to the reset button, we can full power cycle the device but now we need to store the count in flash, storing the count in flash is a problem as we don't want to wear it 

We can't fix 2 but we can't limit it, how? Just don't write too much:
Let's assume default setting of 10 loops to enter safe mode and 5 mins of safe mode time before auto reboot. 
Stuff goes wrong for any reason and device is rebooted, after 10 loops device enters safe mode. 
As soon as this happens the loop counter is reset to zero, this is ok if OTA goes thru or user is actively watching logs and testing something. Otherwise we can make sure the counter does not go to zero until
OTA is done or the 5 mins _during_ safe mode elapses complete. This fixes 1.

If OTA does not happen we still want to attempt a normal boot (this is legacy behavior), we can then reset boot count to zero, this will likely cause another 10 loops and a 5 minute in safe mode, and cycle forever. To minimize flash usage I propose 
a) just decrease boot loop 1, so 1 attempt to clan boot will be tried and then will enter in safe mode again, so two writes every 5 minutes, undetermined count but not many
b) keep increasing boot count, try to boot on odd counts and enter safe mode in even counts. This is same as a but once count reaches 20, after the restar time elapses do not reboot, just reset counter to zero. Flash wear will be preserved, at least until **user** power cycles or resets the device. Without user intervention flash is written 20 times, this can be default
c) disbale auto reboot by timeout if store in flash is enabled, I wouldn't like this as I'd like store in flash be default. 

We cannot store in flash we are going to safe mode permanently (soft brick)
a) that's a breaking change 
b) it's not smart (can be read as it's stupid)
c) when there is life there is hope

a not so common reason this can happen if an attached device to esp is in fault condition which would cause the esp to boot loop, replacing the device might get the thing operational again 

A faulty grid power issues / power supply/ a bad wire / a child playing with a switch etc. can power cycle the device many times with no reason being a fault in the device itself causing the device to self destroy (to the end users POV)

In any case storing in flash should default to false if we are using deep sleep as rebooting is normal device behavior.

**Comment by alexyao2015** [comment]
I agree that 2 is a pretty tricky issue that is hard to fix. 

The potential fix outlined with alternate booting between safe mode and normal mode is a potential solution. The issue I see with this is it's very complex and requires many changes. With each change, there is room for an increased amount of bugs, increased chance for user error, increased work to implement, and increased maintenance. The less code and complexity around safe mode results in lower chances for bugs and higher chances of a user recovering their device. I would take a forced OTA that lets me save my device over an uncaught bug that ultimately kills my device due to complex code anyday. Keep it Super Simple or KISS is important for safe mode where we cannot afford any bugs. (I started implementing the feature and realized how much code this really requires and caught many potential bugs along the way. Ex. I made a PR for OTA setting validation that fixes a potential permanent brick caused by changing a simple config option. Think about the bugs that may come with this as it's a substantially more involved feature)

Ultimately, the main benefit we get is simply we are not permanently booting into safe mode, but we risk bugs causing devices to hard brick permanently rather than soft brick. 

This solution, while breaking, is simplistic, requiring substantially less code. We could also make it not breaking by not enabling it by default, but it's typically meant to be an underlying feature that a user will never notice.

Downsides such as issues with the power supply, child, etc can be avoided through 1) increasing the boot count before permanently entering safe mode. 2) disabling this feature. In any case, here I would argue that if a user does have these problems where the device has such a poor power supply or a switch-crazy child that it constantly reboots 10 or 20 times consecutively without being stable for 1 minute between, they should disable this feature to avoid wearing the flash. Even with the proposed solution of alternating between bootloop, we still cannot protect a user from wearing the flash if they have issues like this and it's beyond our control. All we can do is simply suggest to disable the feature to avoid wearing the flash.

I understand that this solution may not be ideal, but I don't see a better way that it can be done while also staying simple.

**Comment by alexyao2015** [comment]
TLDR of ^
1) I don't feel comfortable making large amounts of changes and being 100% sure there are no bugs
2) The alternative still doesn't fix flash wear issues due to power supply, child, etc. while a brick does.

**Comment by badbadc0ffee** [comment]
> 1. I don't feel comfortable making large amounts of changes and being 100% sure there are no bugs
But you *are* 100% sure that the changes you suggest are bug-free although there are no tests in place?

> 2. The alternative still doesn't fix flash wear issues due to power supply, child, etc. while a brick does.
I don't think that wear that is potentially occuring because of a malfunctioning power supply or a child flicking the switch - literally - every second for a full day are out of the scope of the firmware.

**Comment by alexyao2015** [comment]
Yes, I can be 100% sure there are no bugs because there is effectively a change of 2 lines here (storing in flash and changing bootloop reset). Compare that to the hundreds of lines that would need to be changed for such a feature to be implemented. If there were a bug in that, it's unlikely to be caught as well with or without tests.

**Comment by badbadc0ffee** [comment]
I am looking forward to reviewing another PR with two changed liens of code then.

File: esphome/components/ota/__init__.py [hunk 2]
```diff
@@ -2,24 +2,40 @@
...
     cv.GenerateID(): cv.declare_id(OTAComponent),
     cv.Optional(CONF_SAFE_MODE, default=True): cv.boolean,
     cv.SplitDefault(CONF_PORT, esp8266=8266, esp32=3232): cv.port,
     cv.Optional(CONF_PASSWORD, default=''): cv.string,
-    cv.Optional(CONF_REBOOT_TIMEOUT, default='5min'): cv.positive_time_period_milliseconds,
-    cv.Optional(CONF_NUM_ATTEMPTS, default='10'): cv.positive_not_null_int
-}).extend(cv.COMPONENT_SCHEMA)
+    cv.Optional(CONF_REBOOT_TIMEOUT, default='5min'): cv.All(
+        cv.positive_time_period_milliseconds,
+        cv.Range(max=TimePeriod(milliseconds=4294967295)),
```
Comment (badbadc0ffee) [inline esphome/components/ota/__init__.py:34]: This does not belong here - as I already stated in the other PR - and is also incorrect.

**Review by badbadc0ffee — CHANGES_REQUESTED** [review]

**Comment by badbadc0ffee** [comment]
I'll stop reading this PR now. You can notify me when you get it down to the two lines of code you prefigured before instead of bloating it with additional incorrect changes.

**Comment by alexyao2015** [comment]
Its not bloated? If you actually read the changes, you would see the bulk of it is a result of renaming rtc to prefs due to the change in usage. Thats literally the only reason it has extended lines...

**Comment by alexyao2015** [comment]
To sum things up:
1) I added 1 new config option. Validation is ported from another pr.
2) I pass that config option into a preexisting function
3) The config option configures the prefs such that its stored on flash or in rtc.
4) Added 1 if statement to check if the config option is on or off. 
5) Renamed all instances of rtc to prefs so that it makes sense given the new use case.

File: esphome/components/ota/__init__.py [hunk 3]
```diff
@@ -2,7 +2,8 @@
 import esphome.config_validation as cv
 from esphome.const import (
     CONF_ID, CONF_NUM_ATTEMPTS, CONF_PASSWORD,
-    CONF_PORT, CONF_REBOOT_TIMEOUT, CONF_SAFE_MODE
+    CONF_PORT, CONF_STORE_BOOTLOOPS_IN_FLASH_AND_BRICK, CONF_REBOOT_TIMEOUT,
```
Comment (glmnet) [inline esphome/components/ota/__init__.py:5]: I'll join @badbadc0ffee  and stop following this PR too until I stop reading BRICK, I'll merge the time fix now so you can rebase this

Comment (alexyao2015) [reply esphome/components/ota/__init__.py:5]: I can change the working of brick easily. My question to you is, if I do change said working to say `power_cycle_recovery`, will you accept the PR?

Comment (glmnet) [reply esphome/components/ota/__init__.py:5]: Of course not, I said I will not follow the PR 
We can work it out until it's clean code, not likely buggy and of course will not brick the device permanently. Many of us already agreed on that behavior, If you are not ok with that I'm afraid this is not going to happen.

Comment (alexyao2015) [reply esphome/components/ota/__init__.py:5]: I'm really not sure what to make of this. If I recreate this to do what you mentioned, code complexity increases greatly, resulting in potential bugs due to untested areas. Of course, I will do my best to make it bug free and it will be reviewed and tested, but those tests only cover so many cases. 

Even with a small PR like #1393, there were still bugs that got through the cracks requiring #1457. That was a change of a total of 15 lines. Consider a PR to safe mode with hundreds of lines of change. A bug that slips through a PR that is 15 lines changed will almost definitely slip through one that is hundreds of lines of changes. Considering that this is safe mode, it doesn't make sense to essentially refactor the entire thing to implement such a feature given that we know that safe mode as it is now is tried and true. 

A bug in safe mode bricks the device and likely results in it going to a landfill. A bug in a component or any other non-core component bricks the device temporarily into safe mode where a user can OTA.  

The choice is either:
1) Have a disclaimer on this feature as there is now and default it to off so a user cannot unintentionally harm their device.
2) Implement a new way to handle safe mode and risk tons of bugs in the process and devices going to the landfill. 

I simply cannot understand why this is such a controversial feature given that it's an optional feature that is defaulted to off. Someone that turns it on will know exactly what they are doing because there is very clear documentation surrounding it. 

Maybe the suggestion you had is fit for a complete overhaul of OTA, changing how it currently functions today such that the new functionality is a default on. It's out of scope for the issue I'm trying to solve with this PR.

**Comment by alexyao2015** [comment]
I've updated the description of the PR to provide a summary of the comments.

**Comment by alexyao2015** [comment]
Another simple idea/thought to make this better and stay simple. Instead of permanently bricking the device, it:
1) is set to default enabled, but is default disabled for deep sleep
2) once entering safe mode, it permanently enters safe mode until it can cleanly exit safe mode (only couple lines to change to make this a reality)
 + also prevents fast power cycling from destroying the flash
3) no longer permanently in safe mode
4) resolves issues mentioned for creating this PR

**Comment by alexyao2015** [comment]
Last note here. Lots of other iot devices sold to normal consumers use a power cycle as a way of a complete factory reset. Personally, I have Yeelights which will fully reset after 5 power cycles. It's more or less equivalent to this... 

Those with kids or whoever thinks this would be an issue, shouldn't buy any normal hardware sold to the general public is the consensus here....

**Comment by badbadc0ffee** [comment]
The very big difference is that others don't brick their devices but reset the configuration in a way that an end user can recover it without special tooling. That is in particular also the case for Tasmota where you can bring the device into a state where any smartphone can be used to connect to the device in this recovery state, put in the wifi credentials and other configuration parameters, restart and be ready to go again. That is the _opposite_ of bricking a device.

On the other hand, if you change the Tasmota code - to stick with that example that you have also cited by yourself - you can also easily shoot yourself into your knee there and put in or change code that _bricks_ the device if you break the recovery mode countermeasures that are put into place to save you.

As a general rule of thumb, I'd say that if you use a special tool such as a compiler in a way that leads to a '_bricked_' device, it is just fair that another special tool such as a screwdriver or even soldering iron is needed to recover. If you only use a general purpose tool such as a web browser on your phone (or even just the power switch) and accidentally change the configuration in a way that the device appears to be '_broken_', countermeasures such as a power-cycle-recovery procedures should be in place to be able to recover into a safe state and explicitly _avoid_ the '_brick_'.

As this PR is after all of this lengthy discussion still explicitly aiming at '_bricking_' the device even though that can - as we all agree - also happen accidentally just by regular user interaction such as a child flicking a switch a couple of times too often, it seems like it as no perspective to get merged into ESPHome.

If you feel better with such a development feature in place while working on a configuration for a device that cannot be easily be accessed to reflash when something goes wrong, it is of course a pretty neat idea to bring in some countermeasure to save you from _bricking_ your device. I can perfectly understand that. Go ahead, do that on your branch! That is good practice, I'd say.

But this does not belong into the production software. I'd suggest you close this PR now.

**Comment by alexyao2015** [comment]
@badbadc0ffee Did you read this new comment about the direction for this PR? I think this is a good compromise.
https://github.com/esphome/esphome/pull/1420#issuecomment-758229572

**Comment by badbadc0ffee** [comment]
> @badbadc0ffee Did you read this new comment ...

I did.

**Comment by alexyao2015** [comment]
If you did in fact read it, you would realize your comments about bricking the device do not make sense and are irrelevant.

**Review by badbadc0ffee — CHANGES_REQUESTED** [review]
You are wasting my time. Nothing changed.

**Comment by alexyao2015** [comment]
Here's a potential pitfall to the alternating boots suggestion from @glmnet. Another user in the issues section found a situation where safe mode fails to start at all. Suppose that the boot loop counter was stored in flash like required for that feature, the device would boot loop into eternity and destroy the flash in the process, counting up and resetting with each boot.

This solution protects against that. I'm willing to discuss the alternative I suggested, but haven't received any feedback. The changes I made were to simply rebase this onto the dev version.

**Comment by OttoWinter** [comment]
I don't have time to read through the entire comment history, but still trying to understand what this is for.

Could you answer this question:

1. What concrete problem does this solve? 

2. Has such a problem provenly ever happened in real environments? 

3. And How does this PR solve that?

2. Why is the config option include "brick"? As [Merriam webster's suggests](https://www.merriam-webster.com/dictionary/brick#:~:text=Definition%20of%20brick%20(Entry%202,hacking%2C%20or%20software%20changes)%20%E2%80%A6), bricking is the process of _permanently_ destroying a device (as in not recoverable even after a software fix). That is the opposite of what should happen.

It would interest me when the current logic would _not_ work. Especially the hardware watchdog prevents pretty much all issues.

**Comment by alexyao2015** [comment]
I don't think this really makes sense anymore, given some consideration and after #1393. 

The one major issues with safe mode currently is that it can still fail to start with too many entities in the YAML as in esphome/issues#1723, which I had thought this would fix but now realize that it doesn't. Even though the entities are not being added to `App`, it still fails to load.
